src/main/java/uk/nhs/hee/tis/trainee/forms/service/FormRelocateService.java

```diff
@@ -0,0 +1,182 @@
+        try {
+          updateTargetTraineeInDb(formR, sourceTrainee);
+          if (formR.getLifecycleState() != LifecycleState.DRAFT) {
+            relocateFormInS3(formType, formId, targetTrainee, sourceTrainee);
+          }
+        } catch (Exception ex) {
+          log.error("Fail to roll back: " + ex);
+        }
+        throw new ApplicationException("Fail to relocate FormR to target trainee: " + e.toString());
+      }
+    }
+  }
+
+  private AbstractForm getMoveFormInfoInDb(String formId) {
+    try {
+      Optional<FormRPartA> optionalFormRPartA =
+          formRPartARepository.findById(UUID.fromString(formId));
+      Optional<FormRPartB> optionalFormRPartB =
+          formRPartBRepository.findById(UUID.fromString(formId));
+
+      if (optionalFormRPartA.isPresent()) {
+        return optionalFormRPartA.get();
+      }
+      else if (optionalFormRPartB.isPresent()) {
+        return optionalFormRPartB.get();
+      }
+      return null;
+    } catch (Exception e) {
+      log.error("Fail to get FormR with ID " + formId + ": " + e);
+      throw new ApplicationException("Fail to get FormR with ID " + formId + ": " + e.toString());
+    }
+  }
+
+  private void updateTargetTraineeInDb(AbstractForm abstractForm, String targetTrainee) {
+    abstractForm.setTraineeTisId(targetTrainee);
+
+    if (abstractForm instanceof FormRPartA formRPartA) {
+      formRPartARepository.save(formRPartA);
+    } else if (abstractForm instanceof FormRPartB formRPartB) {
+      formRPartBRepository.save(formRPartB);
+    }
+    log.info("Form R with ID " + abstractForm.getId()
+        + " moved under " + targetTrainee + " in DB ");
+  }
+
+  private void relocateFormInS3(
+      String formType, String formId, String sourceTrainee, String targetTrainee)
+      throws IOException {
+
+    String sourceKey = sourceTrainee + "/forms/" + formType + "/" + formId + ".json";
+    String targetKey = targetTrainee + "/forms/" + formType + "/" + formId + ".json";
+
+    log.info("sourceKey : " + sourceKey);
+    log.info("targetKey : " + targetKey);
+
+    // Get Form R from S3
+    S3Object object = amazonS3.getObject(bucketName, sourceKey);
+
+    // Set target trainee Id to S3 metadata
+    ObjectMetadata metadata = object.getObjectMetadata();
+    metadata.addUserMetadata("traineeid", targetTrainee);
+
+    // Set target trainee Id to S3 content
+    S3ObjectInputStream content = object.getObjectContent();
+    ObjectMapper mapper = new ObjectMapper();
+    Map<String, Object> formr = mapper.readValue(content, Map.class);
+    formr.put("traineeTisId", targetTrainee);
+
+    // Save the form under target trainee
+    final PutObjectRequest request = new PutObjectRequest(
+        bucketName,
+        targetKey,
+        new ByteArrayInputStream(mapper.writeValueAsBytes(formr)),
+        metadata);
+    amazonS3.putObject(request);
+    log.info("Form R in S3 relocated from " + sourceKey + " to " + targetKey);
```

Comment:
Can the save to DB and S3 not re-use the existing `AbstractCloudRepository.save(AbstractForm)`?
I can't see a benefit to pulling the S3 object and having to modify the JSON vs just saving the Abstract form to JSON directly.